File: frontend/src/types/training.ts
/**
 * Training-related types for RL algorithms.
 */

import type { DeviceType, Timestamp, UUID, FilePath } from './common';

/** Supported RL algorithms */
export type Algorithm = 'ppo' | 'sac' | 'td3' | 'ddpg' | 'dqn';

/** Training run status */
export type TrainingStatus =
  | 'pending'
  | 'training'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'stopped';

/** PPO algorithm configuration */
export interface PPOConfig {
  learning_rate: number;
  gamma: number;
  gae_lambda: number;
  clip_range: number;
  clip_range_vf?: number;
  entropy_coef: number;
  value_coef: number;
  max_grad_norm: number;
  n_steps: number;
  n_epochs: number;
  batch_size: number;
  normalize_advantage: boolean;
  target_kl?: number;
  use_sde?: boolean;
  sde_sample_freq?: number;
}

/** SAC algorithm configuration */
export interface SACConfig {
  learning_rate: number;
  gamma: number;
  tau: number;
  buffer_size: number;
  learning_starts: number;
  batch_size: number;
  train_freq: number;
  gradient_steps: number;
  ent_coef: 'auto' | number;
  target_update_interval: number;
  target_entropy?: 'auto' | number;
  use_sde?: boolean;
  sde_sample_freq?: number;
}

/** TD3 algorithm configuration */
export interface TD3Config {
  learning_rate: number;
  gamma: number;
  tau: number;
  buffer_size: number;
  learning_starts: number;
  batch_size: number;
  train_freq: number;
  gradient_steps: number;
  policy_delay: number;
  target_policy_noise: number;
  target_noise_clip: number;
}

/** Network architecture configuration */
export interface NetworkConfig {
  policy_layers: number[];
  value_layers: number[];
  activation: 'relu' | 'tanh' | 'elu' | 'gelu';
  ortho_init: boolean;
  log_std_init?: number;
}

/** Training configuration */
export interface TrainingConfig {
  algorithm: Algorithm;
  ppo?: PPOConfig;
  sac?: SACConfig;
  td3?: TD3Config;
  network?: NetworkConfig;
  total_timesteps?: number;
  device?: DeviceType;
  seed?: number;
  verbose?: number;
}

/** Request to start training */
export interface StartTrainingRequest {
  config: TrainingConfig;
  total_steps: number;
  checkpoint_path?: FilePath;
  log_dir?: FilePath;
  experiment_name?: string;
  tags?: string[];
}

/** Training run response */
export interface TrainingRun {
  id: UUID;
  config: Record<string, unknown>;
  status: TrainingStatus;
  current_step: number;
  total_steps: number;
  metrics_count: number;
  start_time?: Timestamp;
  end_time?: Timestamp;
  error_message?: string;
  checkpoint_path?: FilePath;
  experiment_name?: string;
  tags?: string[];
}

/** Training metrics for a single step */
export interface TrainingMetrics {
  step: number;
  policy_loss?: number;
  value_loss?: number;
  kl?: number;
  entropy?: number;
  explained_variance?: number;
  episode_reward?: number;
  episode_length?: number;
  learning_rate?: number;
  clip_fraction?: number;
  approx_kl?: number;
  loss?: number;
  actor_loss?: number;
  critic_loss?: number;
  ent_coef?: number;
  fps?: number;
  time_elapsed?: number;
  total_timesteps?: number;
  timestamp?: Timestamp;
}

/** Aggregated training statistics */
export interface TrainingStats {
  total_episodes: number;
  total_timesteps: number;
  mean_reward: number;
  std_reward: number;
  min_reward: number;
  max_reward: number;
  mean_episode_length: number;
  success_rate?: number;
  training_time_seconds: number;
}

/** Checkpoint information */
export interface Checkpoint {
  id: UUID;
  run_id: UUID;
  step: number;
  path: FilePath;
  size_bytes: number;
  created_at: Timestamp;
  metrics?: TrainingMetrics;
}

/** Evaluation result */
export interface EvaluationResult {
  n_episodes: number;
  mean_reward: number;
  std_reward: number;
  mean_length: number;
  std_length: number;
  success_rate?: number;
  episodes: EpisodeSummary[];
}

/** Episode summary */
export interface EpisodeSummary {
  episode_id: number;
  reward: number;
  length: number;
  success?: boolean;
  info?: Record<string, unknown>;
}

/** Hyperparameter search configuration */
export interface HyperparameterSearchConfig {
  algorithm: Algorithm;
  n_trials: number;
  n_jobs: number;
  study_name: string;
  search_space: Record<string, SearchSpace>;
  pruner?: 'median' | 'hyperband' | 'none';
  sampler?: 'tpe' | 'random' | 'grid';
}

/** Search space definition */
export interface SearchSpace {
  type: 'float' | 'int' | 'categorical';
  low?: number;
  high?: number;
  log?: boolean;
  choices?: (string | number | boolean)[];
}

/** Trial result */
export interface TrialResult {
  trial_id: number;
  params: Record<string, unknown>;
  value: number;
  state: 'complete' | 'pruned' | 'fail';
  datetime_start: Timestamp;
  datetime_complete?: Timestamp;
}

/** Learning curve data point */
export interface LearningCurvePoint {
  timestep: number;
  mean_reward: number;
  std_reward?: number;
  min_reward?: number;
  max_reward?: number;
}

/** Callback configuration */
export interface CallbackConfig {
  eval_freq?: number;
  n_eval_episodes?: number;
  log_freq?: number;
  save_freq?: number;
  early_stopping?: EarlyStoppingConfig;
}

/** Early stopping configuration */
export interface EarlyStoppingConfig {
  reward_threshold: number;
  n_eval_episodes: number;
  patience?: number;
  min_improvement?: number;
}

/** Wandb/MLflow logging config */
export interface ExperimentLoggerConfig {
  provider: 'wandb' | 'mlflow' | 'tensorboard';
  project?: string;
  entity?: string;
  run_name?: string;
  tags?: string[];
  config?: Record<string, unknown>;
}

File: src/arcs/api/schemas/training.py
"""Training-related schemas."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class PPOConfigSchema(BaseModel):
    """PPO algorithm configuration."""

    lr: float = Field(default=3e-4, gt=0, le=1, description="Learning rate")
    clip_ratio: float = Field(default=0.2, gt=0, lt=1, description="PPO clip ratio")
    gamma: float = Field(default=0.99, gt=0, le=1, description="Discount factor")
    gae_lambda: float = Field(default=0.95, ge=0, le=1, description="GAE lambda")
    entropy_coef: float = Field(default=0.01, ge=0, description="Entropy coefficient")
    vf_coef: float = Field(default=0.5, gt=0, description="Value function coefficient")
    update_epochs: int = Field(default=10, ge=1, le=100, description="PPO update epochs")
    batch_size: int = Field(default=64, ge=1, le=8192, description="Batch size")
    steps_per_update: int = Field(default=2048, ge=64, description="Steps per update")


class SACConfigSchema(BaseModel):
    """SAC algorithm configuration."""

    lr: float = Field(default=3e-4, gt=0, le=1, description="Learning rate")
    gamma: float = Field(default=0.99, gt=0, le=1, description="Discount factor")
    tau: float = Field(default=0.005, gt=0, lt=1, description="Target network update rate")
    alpha: float = Field(default=0.2, gt=0, description="Entropy coefficient")


class EnvConfigSchema(BaseModel):
    """Environment configuration."""

    task: Literal["Reach", "Pick", "Place", "Push"] = Field(default="Reach")
    backend: Literal["dummy", "isaacgym", "mujoco", "pybullet"] = Field(default="dummy")
    num_envs: int = Field(default=1, ge=1, le=4096)
    device: Literal["cpu", "cuda"] = Field(default="cpu")


class TrainingConfigSchema(BaseModel):
    """Complete training configuration."""

    ppo: PPOConfigSchema = Field(default_factory=PPOConfigSchema)
    sac: SACConfigSchema = Field(default_factory=SACConfigSchema)
    env: EnvConfigSchema = Field(default_factory=EnvConfigSchema)
    algorithm: Literal["ppo", "sac"] = Field(default="ppo")


class StartTrainingRequest(BaseModel):
    """Request to start a training run."""

    config: TrainingConfigSchema = Field(default_factory=TrainingConfigSchema)
    total_steps: int = Field(default=100000, ge=1000, description="Total training steps")
    checkpoint_interval: int | None = Field(default=10000, description="Checkpoint save interval")
    log_interval: int = Field(default=100, description="Logging interval")


class TrainingMetricsResponse(BaseModel):
    """Training metrics response."""

    step: int = Field(..., description="Training step")
    policy_loss: float = Field(..., description="Policy loss")
    value_loss: float = Field(..., description="Value function loss")
    kl: float = Field(..., description="KL divergence")
    entropy: float = Field(default=0.0, description="Policy entropy")
    explained_variance: float = Field(default=0.0, description="Explained variance")
    episode_reward: float | None = Field(default=None, description="Episode reward")
    episode_length: int | None = Field(default=None, description="Episode length")
    timestamp: float = Field(..., description="Metric timestamp")


class TrainingRunResponse(BaseModel):
    """Training run response."""

    id: str = Field(..., description="Run ID")
    config: dict[str, Any] = Field(..., description="Training configuration")
    status: Literal["idle", "training", "paused", "completed", "error"] = Field(
        default="idle",
        description="Run status"
    )
    current_step: int = Field(default=0, description="Current training step")
    total_steps: int = Field(..., description="Total training steps")
    metrics_count: int = Field(default=0, description="Number of recorded metrics")
    start_time: float | None = Field(default=None, description="Start timestamp")
    end_time: float | None = Field(default=None, description="End timestamp")
    error_message: str | None = Field(default=None, description="Error message if failed")
    created_at: datetime = Field(default_factory=datetime.utcnow)
